genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileServiceImpl.java

```diff
@@ -0,0 +1,563 @@
+        private final Path relativePath;
+        private final long expectedDataSize;
+        private final GRpcJobFileServiceImpl.DelegatingStreamObserver delegatingStreamObserver;
+        private final StreamObserver<ServerJobFileMessage> responseObserver;
+        private final OutputStream outputStream;
+        private final AtomicBoolean closed = new AtomicBoolean();
+        private long bytesReceived;
+        private final FileBuffer fileBuffer;
+
+        ActiveStreamImpl(
+            final String jobId,
+            final Path relativePath,
+            final long startOffset,
+            final long endOffset,
+            final DelegatingStreamObserver delegatingStreamObserver,
+            final StreamObserver<ServerJobFileMessage> responseObserver,
+            final FileBuffer fileBuffer
+        ) {
+            this.jobId = jobId;
+            this.relativePath = relativePath;
+            this.outputStream = fileBuffer.getOutputStream();
+            this.expectedDataSize = endOffset - startOffset;
+            this.delegatingStreamObserver = delegatingStreamObserver;
+            this.responseObserver = responseObserver;
+            this.fileBuffer = fileBuffer;
+
+            this.delegatingStreamObserver.setDelegate(this);
+
+            final ServerJobFileMessage request = ServerJobFileMessage.newBuilder()
+                .setFileStreamRequest(
+                    FileStreamRequest.newBuilder()
+                        .setRelativePath(relativePath.toString())
+                        .setStartOffset(startOffset)
+                        .setEndOffset(endOffset)
+                        .build()
+                )
+                .build();
+            this.responseObserver.onNext(request);
+
+            log.info(
+                "Activated stream for job id: {} to request file: {} [{},{}]",
+                getJobId(),
+                relativePath,
+                startOffset,
+                endOffset
+            );
+        }
+
+        @Override
+        public void handleMessage(
+            final StreamObserver<AgentJobFileMessage> streamObserver,
+            final AgentJobFileMessage value
+        ) {
+            if (value.getMessageCase() != AgentJobFileMessage.MessageCase.FILE_STREAM_CHUNK) {
+                log.warn("Received unexpected message type: {}", value.getMessageCase().name());
+                this.close();
+            } else {
+                final FileStreamChunk chunkMessage = value.getFileStreamChunk();
+                final byte[] chunkBytes = chunkMessage.getFileChunk().toByteArray();
+                log.debug(
+                    "Received chunk for job: {}, file: {} ({} bytes)",
+                    this.jobId,
+                    this.relativePath,
+                    chunkBytes.length
+                );
+                try {
+                    // Write bytes in output stream
+                    this.outputStream.write(chunkBytes);
```

Comment:
This could definitely be avoided if we're ok buffering entirely in memory.
But if 2 users request a multi-GB file at the same time, we're going to crash.

Spilling to file seems to be the only option to me.

Pipe is exactly the right abstraction, but it's not file-backed.

Reply:
What if we instead of just blindly pushed we apply backpressure so that it can fit in whatever pre-allocated buffers we allow? Then we wouldn't have to write to disk

Reply:
Certainly doable. A lot more complicated because there are 2 streams: agent-server and server-client.
The backpressure from server-client needs to be applied to the agent-server stream.

The server-client stream does not provide proper back-pressure signal, we can guess from what they are reading and make some assumptions and use the result to apply backpressure.
But it will make the two a lot more complex and tightly coupled.

On top of that, we need to keep into account all other streams on the same server since they're all sharing memory.

I'm not a big fan of using the JVM heap for transient data like this. 
We need to reserve a big chunk of it to these buffers if we want to allow a decent number of parallel downloads.
Do we have all of that spare heap space to begin with?